ruby/red-arrow/ext/arrow/record_batch.cc

```diff
@@ -0,0 +1,610 @@
+
+#undef CONVERT_SIGNED_INTEGER
+
+  template <typename ArrayType>
+  inline VALUE ConvertUnsignedIntegerValue(const ArrayType& array, const int64_t i) {
+    return rb::protect([&]{ return ULL2NUM(array.Value(i)); });
+  }
+
+#define CONVERT_UNSIGNED_INTEGER(TYPE) \
+  VALUE ConvertValue(const TYPE& array, const int64_t i) { \
+    return ConvertUnsignedIntegerValue<TYPE>(array, i); \
+  }
+
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt8Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt16Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt32Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt64Array)
+
+#undef CONVERT_UNSIGNED_INTEGER
+
+  template <typename ArrayType>
+  inline VALUE ConvertFloatValue(const ArrayType& array, const int64_t i) {
+    return rb::protect([&]{ return DBL2NUM(array.Value(i)); });
+  }
+
+#define CONVERT_FLOAT(TYPE) \
+  VALUE ConvertValue(const TYPE& array, const int64_t i) { \
+    return ConvertFloatValue<TYPE>(array, i); \
+  }
+
+  CONVERT_FLOAT(arrow::FloatArray)
+  CONVERT_FLOAT(arrow::DoubleArray)
+
+#undef CONVERT_FLOAT
+
+  inline VALUE ConvertValue(const arrow::HalfFloatArray& array, const int64_t i) {
+    // FIXME: should convert to Float
+    return rb::protect([&]{
+      return LONG2FIX(static_cast<long>(array.Value(i)));
+    });
+  }
+
+  inline VALUE ConvertDecimal128(const arrow::Decimal128Array& array, const int64_t i) {
+    return rb::protect([&]{
+      auto raw_value = std::make_shared<arrow::Decimal128>(array.GetValue(i));
+      auto gobj_value = garrow_decimal128_new_raw(&raw_value);
+      return GOBJ2RVAL(gobj_value);
+    });
+  }
+
+  inline VALUE ConvertDecimal128ToBigDecimal(const arrow::Decimal128Array& array, const int64_t i) {
+    auto decimal_str = array.FormatValue(i);
+    return rb::protect([&]{
+      return rb_funcall(rb_cObject, id_BigDecimal, 1, rb_str_new_cstr(decimal_str.c_str()));
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Decimal128Array& array, const int64_t i) {
+    if (convert_decimal_) {
+      return ConvertDecimal128ToBigDecimal(array, i);
+    } else {
+      return ConvertDecimal128(array, i);
+    }
+  }
+
+  inline VALUE ConvertValue(const arrow::FixedSizeBinaryArray& array, const int64_t i, VALUE buffer) {
+    const auto byte_width = array.byte_width();
+    return rb::protect([&]{ return rb_str_substr(buffer, i*byte_width, byte_width); });
+  }
+
+  inline VALUE ConvertValue(const arrow::FixedSizeBinaryArray& array, const int64_t i) {
+    const auto byte_width = array.byte_width();
+    const auto offset = i * byte_width;
+    const auto ptr = reinterpret_cast<const char*>(array.raw_values());
+    return rb::protect([&]{ return rb_str_new(ptr + offset, byte_width); });
+  }
+
+  inline VALUE ConvertValue(const arrow::BinaryArray& array, const int64_t i) {
+    int32_t length;
+    const uint8_t* ptr = array.GetValue(i, &length);
+    return rb::protect([&]{
+      return rb_str_new(reinterpret_cast<const char*>(ptr), length);
```

Comment:
Could we use `rb_enc_str_new(ptr, length, rb_ascii8bit_encoding)` here?

Reply:
Replaced.